File: backend/TRFSAE.MemberPortal.API/Enums.cs
using System.Runtime.Serialization;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace TRFSAE.MemberPortal.API.Enums;

[JsonConverter(typeof(StringEnumConverter))]
public enum TRSystem
{
    Chassis,
    Powertrain,
    Software,
    Business
}

[JsonConverter(typeof(StringEnumConverter))]
public enum Subsystem
{
    // Chassis
    Frame,
    Aerodynamics,
    Ergonomics,
    Brakes,
    Suspension,

    // Powertrain
    Battery,
    Electronics,
    LowVoltage,
    TractiveSystem,

    // Software
    App,
    Embedded,
    Data,

    // Business
    Finance,
    PublicRelations
}

[JsonConverter(typeof(StringEnumConverter))]
public enum ShirtSize
{
    XS,
    S,
    M,
    L,
    XL,
    XXL
}

[JsonConverter(typeof(StringEnumConverter))]
public enum Role
{
    SuperAdmin,
    Admin,
    SystemLead,
    SubsystemLead,
    Member
}

[JsonConverter(typeof(StringEnumConverter))]
public enum ProjectPriority
{
    Low,
    Medium,
    High
}

[JsonConverter(typeof(StringEnumConverter))]
public enum OrderStatus
{
    Pending,
    Denied,
    Approved,
    Delivering,
    Delivered,
    Claimed
}

File: shared/config/enums.ts
/**
 * Enums shared between frontend and database.
 * Note: PostgreSQL requires that all enum members are of type string.
 */

export enum System {
    Chassis = "Chassis",
    Powertrain = "Powertrain",
    Software = "Software",
    Business = "Business"
}

export enum Subsystem {
    // Chassis
    Frame = "Frame",
    Aerodynamics = "Aerodynamics",
    Ergonomics = "Ergonomics",
    Brakes = "Brakes",
    Suspension = "Suspension",

    // Powertrain
    Battery = "Battery",
    Electronics = "Electronics",
    LowVoltage = "Low Voltage",
    TractiveSystem = "Tractive System",

    // Software
    App = "App Development",
    Embedded = "Embedded Systems",
    Data = "Data Acquisition",

    // Business
    PublicRelations = "Public Relations",
    Finance = "Finance"
}

export enum ShirtSize {
    XS = "Extra Small",
    S = "Small",
    M = "Medium",
    L = "Large",
    XL = "Extra Large",
    XXL = "Extra Extra Large"
}

export enum Role {
    SuperAdmin = "Superadmin",
    Admin = "Admin",
    SystemLead = "System Lead",
    SubsystemLead = "Subsystem Lead",
    Member = "Member"
}

export enum ProjectPriority {
    Low = "Low",
    Medium = "Medium",
    High = "High"
}

export enum OrderStatus {
    Pending = "pending",
    Denied = "denied",
    Approved = "approved",
    Delivering = "delivering",
    Delivered = "delivered",
    Claimed = "claimed"
}
